File: app/src/main/java/com/agrogames/islandsofwar/engine/impl/navigator/Direction.java
package com.agrogames.islandsofwar.engine.impl.navigator;

import com.agrogames.islandsofwar.engine.abs.common.Cell;

enum Direction {
    Top, Right, Bottom, Left;

    /**
     * goal != current
     * @return if goal==current, BOTTOM
     */
    public static Direction primary(Cell current, Cell goal) {
        if (module(goal.x - current.x) > module(goal.y - current.y)) {
            if (goal.x > current.x) {
                return Direction.Right;
            } else {
                return Direction.Left;
            }
        } else {
            if (goal.y > current.y) {
                return Direction.Top;
            } else {
                return Direction.Bottom;
            }
        }
    }

    /**
     * goal != current
     * @return if goal==current, BOTTOM
     */
    public static Direction secondary(Cell current, Cell goal) {
        if (module(goal.x - current.x) < module(goal.y - current.y)) {
            if (goal.x > current.x) {
                return Direction.Right;
            } else {
                return Direction.Left;
            }
        } else {
            if (goal.y > current.y) {
                return Direction.Top;
            } else {
                return Direction.Bottom;
            }
        }
    }

    /**
     * goal != current
     * @return if goal==current, BOTTOM
     */
    public static Direction tertiary(Cell current, Cell goal) {
        if (module(goal.x - current.x) < module(goal.y - current.y)) {
            if (goal.x <= current.x) {
                return Direction.Right;
            } else {
                return Direction.Left;
            }
        } else {
            if (goal.y <= current.y) {
                return Direction.Top;
            } else {
                return Direction.Bottom;
            }
        }
    }

    /**
     * goal != current
     * @return if goal==current, BOTTOM
     */
    public static Direction quaternary(Cell current, Cell goal) {
        if (module(goal.x - current.x) > module(goal.y - current.y)) {
            if (goal.x <= current.x) {
                return Direction.Right;
            } else {
                return Direction.Left;
            }
        } else {
            if (goal.y <= current.y) {
                return Direction.Top;
            } else {
                return Direction.Bottom;
            }
        }
    }

    private static int module(int a) {
        return a >= 0 ? a : -a;
    }
}

File: app/src/main/java/com/agrogames/islandsofwar/engine/impl/navigator/Direction.kt
package com.agrogames.islandsofwar.engine.impl.navigator

import com.agrogames.islandsofwar.engine.abs.common.Cell

enum class Direction {
    Top, Right, Bottom, Left;

    companion object {
        /**
         * goal != current
         * @return if goal==current, BOTTOM
         */
        fun primary(current: Cell?, goal: Cell?): Direction {
            return if (module(goal!!.x - current!!.x) > module(goal.y - current.y)) {
                if (goal.x > current.x) {
                    Right
                } else {
                    Left
                }
            } else {
                if (goal.y > current.y) {
                    Top
                } else {
                    Bottom
                }
            }
        }

        /**
         * goal != current
         * @return if goal==current, BOTTOM
         */
        fun secondary(current: Cell?, goal: Cell?): Direction {
            return if (module(goal!!.x - current!!.x) < module(goal.y - current.y)) {
                if (goal.x > current.x) {
                    Right
                } else {
                    Left
                }
            } else {
                if (goal.y > current.y) {
                    Top
                } else {
                    Bottom
                }
            }
        }

        /**
         * goal != current
         * @return if goal==current, BOTTOM
         */
        fun tertiary(current: Cell?, goal: Cell?): Direction {
            return if (module(goal!!.x - current!!.x) < module(goal.y - current.y)) {
                if (goal.x <= current.x) {
                    Right
                } else {
                    Left
                }
            } else {
                if (goal.y <= current.y) {
                    Top
                } else {
                    Bottom
                }
            }
        }

        /**
         * goal != current
         * @return if goal==current, BOTTOM
         */
        fun quaternary(current: Cell?, goal: Cell?): Direction {
            return if (module(goal!!.x - current!!.x) > module(goal.y - current.y)) {
                if (goal.x <= current.x) {
                    Right
                } else {
                    Left
                }
            } else {
                if (goal.y <= current.y) {
                    Top
                } else {
                    Bottom
                }
            }
        }

        private fun module(a: Int): Int {
            return if (a >= 0) a else -a
        }
    }
}